### ISSUE TITLE
systemctl start rapiddiskd fails on RockyLinux 8.5

### ISSUE THREAD

**Issue by q5sys** [pkoutoupis/rapiddisk#88]
The install itself went fine once I got the deps figured out.  The build compiled as it should, the module was loaded without issue, but when I try to start the service it fails, and doesn't really give a reason.

```
[q5sys@localhost rapiddisk]$ sudo systemctl start rapiddiskd.service
Job for rapiddiskd.service failed because the control process exited with error code.
See "systemctl status rapiddiskd.service" and "journalctl -xe" for details.

[q5sys@localhost rapiddisk]$ systemctl status rapiddiskd.service
● rapiddiskd.service - A rapiddisk network management daemon
   Loaded: loaded (/etc/systemd/system/rapiddiskd.service; disabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Sun 2022-05-01 12:40:15 EDT; 1min 31s ago
  Process: 629839 ExecStart=/sbin/rapiddiskd (code=exited, status=127)

May 01 12:40:15 localhost.localdomain systemd[1]: rapiddiskd.service: Service RestartSec=100ms expired, scheduling restart.
May 01 12:40:15 localhost.localdomain systemd[1]: rapiddiskd.service: Scheduled restart job, restart counter is at 5.
May 01 12:40:15 localhost.localdomain systemd[1]: Stopped A rapiddisk network management daemon.
May 01 12:40:15 localhost.localdomain systemd[1]: rapiddiskd.service: Start request repeated too quickly.
May 01 12:40:15 localhost.localdomain systemd[1]: rapiddiskd.service: Failed with result 'exit-code'.
May 01 12:40:15 localhost.localdomain systemd[1]: Failed to start A rapiddisk network management daemon.

[q5sys@localhost rapiddisk]$ journalctl -xe
May 01 12:54:08 localhost.localdomain systemd[1]: rapiddiskd.service: Failed with result 'exit-code'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- The unit rapiddiskd.service has entered the 'failed' state with result 'exit-code'.
May 01 12:54:08 localhost.localdomain systemd[1]: Failed to start A rapiddisk network management daemon.
-- Subject: Unit rapiddiskd.service has failed
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- Unit rapiddiskd.service has failed.
--
-- The result is failed.
May 01 12:54:09 localhost.localdomain systemd[1]: rapiddiskd.service: Service RestartSec=100ms expired, scheduling restart.
May 01 12:54:09 localhost.localdomain systemd[1]: rapiddiskd.service: Scheduled restart job, restart counter is at 5.
-- Subject: Automatic restarting of a unit has been scheduled
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- Automatic restarting of the unit rapiddiskd.service has been scheduled, as the result for
-- the configured Restart= setting for the unit.
May 01 12:54:09 localhost.localdomain systemd[1]: Stopped A rapiddisk network management daemon.
-- Subject: Unit rapiddiskd.service has finished shutting down
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
--
-- Unit rapiddiskd.service has finished shutting down.
```

Did I miss something in the docs?  Does the service file need to be configured before starting it?

**Comment by pkoutoupis** [comment]
@q5sys Within the next few days I will install Rocky 8.5 and attempt to debug if it is a bug. Thank you for filing this.

**Comment by q5sys** [comment]
@pkoutoupis Happy to help anyway I can.  You're project is exactly what I've been looking for, for quite a while.  I don't mind using another Linux distro if I need to.  I just already had a Rocky machine up and running so it seemed like a good fit.  Do you have a list of distros that Rapiddisk has been confirmed to work on?

**Comment by pkoutoupis** [comment]
@q5sys Thank you.

> Note: The rapiddisk compile would fail as it couldn't find microhttpd.h even though libmicrohttpd was installed.

That is because you need to install `epel-release` and:
```
$ sudo dnf config-manager --set-enabled powertools
```
Followed by:
```
$ dnf install libmicrohttpd-devel
```

Doing that, I had no issues:
```
[root@localhost rapiddisk]# systemctl start rapiddiskd.service
[root@localhost rapiddisk]# systemctl status rapiddiskd.service
● rapiddiskd.service - A rapiddisk network management daemon
   Loaded: loaded (/etc/systemd/system/rapiddiskd.service; enabled; vendor preset: disabled)
   Active: active (running) since Mon 2022-05-02 15:16:01 EDT; 5s ago
  Process: 21300 ExecStart=/sbin/rapiddiskd (code=exited, status=0/SUCCESS)
 Main PID: 21301 (rapiddiskd)
    Tasks: 3 (limit: 11407)
   Memory: 796.0K
   CGroup: /system.slice/rapiddiskd.service
           └─21301 /sbin/rapiddiskd

May 02 15:16:01 localhost.localdomain systemd[1]: Starting A rapiddisk network management daemon...
May 02 15:16:01 localhost.localdomain rapiddiskd[21300]: rapiddiskd 8.1.0
May 02 15:16:01 localhost.localdomain rapiddiskd[21300]: Copyright 2011 - 2022 Petros Koutoupis
May 02 15:16:01 localhost.localdomain systemd[1]: Started A rapiddisk network management daemon.
May 02 15:16:01 localhost.localdomain rapiddiskd[21301]: rapiddiskd: Daemon started.
[root@localhost rapiddisk]# systemctl status rapiddiskd.service
● rapiddiskd.service - A rapiddisk network management daemon
   Loaded: loaded (/etc/systemd/system/rapiddiskd.service; enabled; vendor preset: disabled)
   Active: active (running) since Mon 2022-05-02 15:16:01 EDT; 10s ago
  Process: 21300 ExecStart=/sbin/rapiddiskd (code=exited, status=0/SUCCESS)
 Main PID: 21301 (rapiddiskd)
    Tasks: 3 (limit: 11407)
   Memory: 796.0K
   CGroup: /system.slice/rapiddiskd.service
           └─21301 /sbin/rapiddiskd
```

And `ps` also confirms that it is running:
```
# ps aux|grep rapiddisk
root       21301  0.0  0.0 184940   512 ?        Ssl  15:16   0:00 /sbin/rapiddiskd
root       21314  0.0  0.0 221928  1036 pts/0    R+   15:20   0:00 grep --color=auto rapiddisk
```

The daemon is also responsive:
```
[root@localhost rapiddisk]# curl -s --output - 127.0.0.1:9118/v1/checkServiceStatus|jq .
{
  "status": "OK",
  "version": "8.1.0"
}
[root@localhost rapiddisk]# rapiddisk -a 128
rapiddisk 8.1.0
Copyright 2011 - 2022 Petros Koutoupis

Attached device rd0 of size 128 Mbytes
[root@localhost rapiddisk]# curl -s --output - 127.0.0.1:9118/v1/listRapidDiskVolumes|jq .
{
  "volumes": [
    {
      "rapiddisk": [
        {
          "device": "rd0",
          "size": 134217728
        }
      ]
    },
    {
      "rapiddisk_cache": []
    }
  ]
}
```

Oh, and....
```
[root@localhost rapiddisk]# uname -a
Linux localhost.localdomain 4.18.0-348.23.1.el8_5.x86_64 #1 SMP Wed Apr 27 15:32:52 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
[root@localhost rapiddisk]# cat /etc/redhat-release 
Rocky Linux release 8.5 (Green Obsidian)
```

I would retry with the Powertools repo enabled version of libmicrohttpd-devel, to make sure that your header matches the installed library and we can then work from there. Also, after that, see if you can manually load it with `/sbin/rapiddiskd`. It will fork and bring you back to the prompt. Does it run (verify with ps)?

**Comment by q5sys** [comment]
Hmm.... I've got some free time tomorrow night I'll try it again.  I have EPEL and the powertools repo enabled... but I might have had the order wrong to get the right packages.
If I can reproduce your results I'll close the issue.  Thanks for investigating.

**Comment by matteotenca** [comment]
I suggest a change to the service file:

```diff
--- conf/rapiddiskd.service     2022-05-05 00:28:34.802209912 +0200
+++ conf/rapiddiskd.service   2022-05-04 23:02:53.597818531 +0200
@@ -3,9 +3,12 @@
 After=network.target

 [Service]
+ExecStartPre=sh -c 'lsmod | grep -q rapiddisk || { modprobe rapiddisk ; modprobe rapiddisk-cache ; }'
 ExecStart=/sbin/rapiddiskd
 Type=forking
 Restart=always
+RestartPreventExitStatus=255

```

This way the modules are always loaded before `rapiddiskd` is executed, and when `rapiddiskd` fails with a return value of 255 (this happend to me when modules were not loaded yet), the system stops trying to restart it forever.

**Comment by pkoutoupis** [comment]
Awesome suggestions. Thank you.

**Comment by q5sys** [comment]
@pkoutoupis I figured out the issue.  I did a clean install and was still running into problems.  Fired up strace and tried to start the daemon manually and immediately realized where I messed up.  I was only loading the main module and not the cache module.  Since I had no plans to use rapiddisk as a disk cache I didn't think it was required.
The main readme shows loading/unloading the modules individually so I wrongly assumed that the two modules were independent of each other. 

I'm closing out this ticket as I can load the service fine now. Thanks for your assistance!

I've put in a PR with @matteotenca 's suggestion as well as a few lines so journalctl logs stderr.  AAlso I added a note to the main readme file about both modules needing to be loaded.
https://github.com/pkoutoupis/rapiddisk/pull/92
